From c86a755419658e6e7ebcaaa691050c54ae5c97e7 Mon Sep 17 00:00:00 2001
From: Yonghua Zheng <yh.zheng@yuneec.com>
Date: Fri, 14 Sep 2018 09:40:06 +0800
Subject: [PATCH 1/6] Add custom build variant

Unlike [debug|userdebug|eng] that could introduce addtional packages,
custom variant gives device a full control over what to be built.
---
 core/envsetup.mk       | 4 ++--
 core/product_config.mk | 2 +-
 envsetup.sh            | 8 ++++----
 3 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/core/envsetup.mk b/core/envsetup.mk
index 8ac437e..6dcf3ec 100644
--- a/core/envsetup.mk
+++ b/core/envsetup.mk
@@ -114,10 +114,10 @@ TARGET_COPY_OUT_RECOVERY := recovery
 # variables that we need in order to locate the output files.
 include $(BUILD_SYSTEM)/product_config.mk
 
-build_variant := $(filter-out eng user userdebug,$(TARGET_BUILD_VARIANT))
+build_variant := $(filter-out eng user userdebug custom,$(TARGET_BUILD_VARIANT))
 ifneq ($(build_variant)-$(words $(TARGET_BUILD_VARIANT)),-1)
 $(warning bad TARGET_BUILD_VARIANT: $(TARGET_BUILD_VARIANT))
-$(error must be empty or one of: eng user userdebug)
+$(error must be empty or one of: eng user userdebug custom)
 endif
 
 # ---------------------------------------------------------------
diff --git a/core/product_config.mk b/core/product_config.mk
index e4b68cf..3a19176 100644
--- a/core/product_config.mk
+++ b/core/product_config.mk
@@ -82,7 +82,7 @@ endef
 # These are the valid values of TARGET_BUILD_VARIANT.  Also, if anything else is passed
 # as the variant in the PRODUCT-$TARGET_BUILD_PRODUCT-$TARGET_BUILD_VARIANT form,
 # it will be treated as a goal, and the eng variant will be used.
-INTERNAL_VALID_VARIANTS := user userdebug eng
+INTERNAL_VALID_VARIANTS := user userdebug eng custom
 
 # ---------------------------------------------------------------
 # Provide "PRODUCT-<prodname>-<goal>" targets, which lets you build
diff --git a/envsetup.sh b/envsetup.sh
index 0661fe6..cf5fcc6 100644
--- a/envsetup.sh
+++ b/envsetup.sh
@@ -2,7 +2,7 @@ function hmm() {
 cat <<EOF
 Invoke ". build/envsetup.sh" from your shell to add the following functions to your environment:
 - lunch:   lunch <product_name>-<build_variant>
-- tapas:   tapas [<App1> <App2> ...] [arm|x86|mips|armv5] [eng|userdebug|user]
+- tapas:   tapas [<App1> <App2> ...] [arm|x86|mips|armv5] [eng|userdebug|user|custom]
 - croot:   Changes directory to the top of the tree.
 - m:       Makes from the top of the tree.
 - mm:      Builds all of the modules in the current directory, but not their dependencies.
@@ -66,7 +66,7 @@ function check_product()
     # hide successful answers, but allow the errors to show
 }
 
-VARIANT_CHOICES=(user userdebug eng)
+VARIANT_CHOICES=(user userdebug eng custom)
 
 # check to see if the supplied variant is valid
 function check_variant()
@@ -541,8 +541,8 @@ complete -F _lunch lunch
 function tapas()
 {
     local arch=$(echo -n $(echo $* | xargs -n 1 echo | \grep -E '^(arm|x86|mips|armv5)$'))
-    local variant=$(echo -n $(echo $* | xargs -n 1 echo | \grep -E '^(user|userdebug|eng)$'))
-    local apps=$(echo -n $(echo $* | xargs -n 1 echo | \grep -E -v '^(user|userdebug|eng|arm|x86|mips|armv5)$'))
+    local variant=$(echo -n $(echo $* | xargs -n 1 echo | \grep -E '^(user|userdebug|eng|custom)$'))
+    local apps=$(echo -n $(echo $* | xargs -n 1 echo | \grep -E -v '^(user|userdebug|eng|custom|arm|x86|mips|armv5)$'))
 
     if [ $(echo $arch | wc -w) -gt 1 ]; then
         echo "tapas: Error: Multiple build archs supplied: $arch"
-- 
2.7.4

